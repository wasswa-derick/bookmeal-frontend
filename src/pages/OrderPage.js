import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import Loader from "react-loader";
import { getCartOrder, postOrder } from "../actions/orders";
import { setMessage } from "../actions/message";

/**
 * @export
 * @class OrderPage
 * @extends {React.Component}
 */
export class OrderPage extends React.Component {
  state = {
    loaded: false
  };
  componentWillMount = () => {
    this.props.getCartOrder();
    this.setState({ loaded: true });
  };
  cancelOrder = () => {
    sessionStorage.removeItem("cartOrder");
    this.props.history.push("/");
  };
  makeOrder = () => {
    const { meals, menuId, orderCount } = this.props.cartOrder;
    this.props
      .postOrder({ meals: meals.map(m => m.id), menuId, orderCount })
      .then(() => {
        sessionStorage.removeItem("cartOrder");
        this.props.setMessage({
          text: `Your order has been placed. thank you`,
          type: "info"
        });
        this.props.history.push("/orders");
      })
      .catch(err => console.log(err));
  };

  /**
   * @returns {div} renders a div
   * @memberof OrderPage
   */
  render() {
    const { meals, cost, orderCount, totalCost } = this.props.cartOrder;
    const { loaded } = this.state;
    return (
      <div className="container">
        {meals.length === 0 ? (
          <div>
            <h4>Empty Order Cart</h4>
          </div>
        ) : (
          <div>
            <h4>Confirm Meals Order:</h4>
          </div>
        )}
        <hr />
        <Loader loaded={loaded}>
          <table className="table table-striped">
            <thead>
              <tr>
                <th>Meal Title</th>
                <th>Meal Description</th>
                <th>Meal Price</th>
              </tr>
            </thead>

            <tbody>
              {meals.map(meal => (
                <tr key={meal.id}>
                  <td>{meal.title}</td>
                  <td>
                    <p>{meal.description}</p>
                  </td>
                  <td>
                    <label htmlFor="price">UGX {meal.price}</label>
                  </td>
                  {/* <td>
                    <i className="fa fa-remove" />
                  </td> */}
                </tr>
              ))}
              <tr>
                <td />
                <td>
                  <label style={{ fontWeight: "bold" }} htmlFor="total">
                    {orderCount} Orders X {totalCost}/= :
                  </label>
                  <br />
                  <label style={{ color: " #00a1e0" }} htmlFor="cost">
                    {" "}
                    Total: UGX {cost}
                  </label>
                </td>
                <td>
                  <button
                    onClick={this.cancelOrder}
                    className="btn btn-danger mr-4"
                  >
                    Cancel
                  </button>
                  <button onClick={this.makeOrder} className="btn btn-success">
                    Confirm
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </Loader>
      </div>
    );
  }
}

OrderPage.propTypes = {
  cartOrder: PropTypes.shape({
    menuId: PropTypes.number.isRequired,
    meals: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        price: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired
      }).isRequired
    ).isRequired,
    totalCost: PropTypes.number.isRequired,
    orderCount: PropTypes.number.isRequired,
    cost: PropTypes.number.isRequired
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  getCartOrder: PropTypes.func.isRequired,
  postOrder: PropTypes.func.isRequired,
  setMessage: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  cartOrder: state.ordersReducer.cartOrder
});

export default connect(mapStateToProps, {
  getCartOrder,
  postOrder,
  setMessage
})(OrderPage);
