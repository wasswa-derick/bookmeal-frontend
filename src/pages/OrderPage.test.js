import React from "react";
import { shallow } from "enzyme";
import { OrderPage } from "./OrderPage";

const err = {
  response: {
    status: 400
  }
};
const fn = () => Promise.reject(err);

describe("OrderPage", () => {
  let wrapper;
  beforeEach(() => {
    const history = {
      push: jest.fn
    };
    const cartOrder = {
      menuId: 1,
      meals: [{ title: "meal lorem", id: 1, price: 1000, description: "desc" }],
      totalCost: 1000,
      orderCount: 1,
      cost: 1000
    };
    wrapper = shallow(
      <OrderPage
        cartOrder={cartOrder}
        history={history}
        setMessage={jest.fn}
        postOrder={fn}
        getCartOrder={fn}
      />
    );
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
