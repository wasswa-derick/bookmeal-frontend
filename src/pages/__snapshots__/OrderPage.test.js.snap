// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderPage should render correctly 1`] = `
<div
  className="container"
>
  <div>
    <h4>
      Confirm Meals Order:
    </h4>
  </div>
  <hr />
  <Component
    component="div"
    loaded={true}
    loadedClassName="loadedContent"
    parentClassName="loader"
  >
    <table
      className="table table-striped"
    >
      <thead>
        <tr>
          <th>
            Meal Title
          </th>
          <th>
            Meal Description
          </th>
          <th>
            Meal Price
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="1"
        >
          <td>
            meal lorem
          </td>
          <td>
            <p>
              desc
            </p>
          </td>
          <td>
            <label
              htmlFor="price"
            >
              UGX 
              1000
            </label>
          </td>
        </tr>
        <tr>
          <td />
          <td>
            <label
              htmlFor="total"
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
            >
              1
               Orders X 
              1000
              /= :
            </label>
            <br />
            <label
              htmlFor="cost"
              style={
                Object {
                  "color": " #00a1e0",
                }
              }
            >
               
              Total: UGX 
              1000
            </label>
          </td>
          <td>
            <button
              className="btn btn-danger mr-4"
              onClick={[Function]}
            >
              Cancel
            </button>
            <button
              className="btn btn-success"
              onClick={[Function]}
            >
              Confirm
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </Component>
</div>
`;
